### ISSUE TITLE
Custom shouldTrigger for SignatureHelpProvider

### ISSUE THREAD

**Issue by NoahTheDuke** [neoclide/coc.nvim#3834]
**Is your feature request related to a problem? Please describe.**
I develop [coc-clojure](https://github.com/NoahTheDuke/coc-clojure). Right now, SignatureManager is created by the LanguageClient and the method `shouldTrigger` can't be changed. I would like to do slightly more complicated logic when determining if I should display the Signature Help window.

**Describe the solution you'd like**
I would like `SignatureManager` to use the chosen SignatureHelpProvider's `shouldTrigger` method if it exists, or default to the current implementation (checking if the character is in the chosen SignatureHelpProvider's `triggerCharacters`).

**Describe alternatives you've considered**
Instead of a `triggerCharacters` list, maybe SignatureHelpProvider could have a `triggerAny` flag that skips the current `triggercharacter` check. Another idea is providing a sentinel value to `triggerCharacters` list to bypass the check.

I am willing to write up an PR for this if desired.

**Comment by chemzqm** [comment]
It's used for auto trigger, you can trigger signature help by `editor.action.triggerParameterHints`:
``` typescript
  events.on('TextInsert', async (bufnr, info, character) => {
    if (something) {
      await commands.executeCommand('editor.action.triggerParameterHints')
    }
  }, null, this.disposables)
```

**Comment by NoahTheDuke** [comment]
Oh that's excellent, thank you.

**Comment by NoahTheDuke** [comment]
I've been experimenting with this and it's not quite what I'm looking for. Every time `commands.executeCommamd` is called, it sets an undo point, as if I had exited and re-entered Insert mode.

This means that if I press `i` to enter Insert mode, type a complete function call, press `Esc` to exit Insert mode, and then press `u` to undo the whole call, it only goes back to the previous character I typed.

Is there a way to trigger `signatureHelp` that doesn't set an undo point?
